add originalId to DplaMapModel

AudreyAltman opened the pull request:
This adds `originalId` to `DplaMapModel` and implements the mapping for all providers.

I did not find a way to apply a single mapping from `getProviderId` to `originalId` that would be applied to all providers.  This is because `getProviderId` belongs to the `IdMinter` trait and `originalId` belongs to the `Mapping` trait.  The only classes with access to methods from both traits are the individual provider mappings.  If anyone has anyone has another idea about this, please let me know.

I tested this locally by running a mapping over a Digital Viriginias harvest, loading the mapped avro files into spark shell, and checking that the values in `originalRecord` were present and seemed correct.

This addresses tickets, DT-760, DT-762, and DT-763.

mdellabitta commented on src/main/scala/dpla/ingestion3/model/ModelConverter.scala:
```diff
@@ -32,7 +32,8 @@ object ModelConverter {
     sidecar = optionalJValue(row, 11),
-    messages = toMulti(row, 12, toIngestMessage)
+    messages = toMulti(row, 12, toIngestMessage),
+    originalId = optionalString(row, 13)
```
Like
`optionalString(row, 13).getOrElse("MISSING")`

AudreyAltman replied:
Ah I see.  Thanks for the clarification.

moltude commented on src/main/scala/dpla/ingestion3/mappers/utils/Mapping.scala:
```diff
@@ -12,6 +12,7 @@ trait Mapping[T] {
 
   // OreAggregation
   def dplaUri(data: Document[T]): ExactlyOne[URI]
+  def originalId(data: Document[T]): ZeroToOne[String] = None
```
Addressing two comments here: 

If we merge IdMinter and Mapping traits [which I don't object to] we would handle this in the same way we handle `dplaUri`, that is to say throw a RuntimeException if no value exists. The IdMinter methods do not record error messages the same way that Mapping methods do. We could either leave this as is and have all the id-related errors show up as unclassified Exceptions for the time being OR implement a validation method for `originalId` in the same way we are mapping the required property `dataProvider/provider`. 

My preference would be to eliminate the RuntimeExceptions by merging IdMinter and Mapping into a single trait. Making the cardinality of `originalId` `ExactlyOne` and inserting `validateOriginalId` code to `Mapper` using `dataProvider` validation as a template. 

@AudreyAltman @mdellabitta thoughts?

mdellabitta replied:
I like it. RuntimeExceptions are really not fashion-forward.

AudreyAltman replied:
Ok, I will work through these changes and let you know when its ready to be reviewed again.

moltude commented on src/main/scala/dpla/ingestion3/mappers/Mapper.scala:
```diff
@@ -316,7 +317,8 @@ class JsonMapper extends Mapper[JValue, JsonMapping] {
         ),
-        messages = msgCollector.getAll()
+        messages = msgCollector.getAll(),
+        originalId = mapping.originalId(document)
```
swapping lines 320 and 321 is the only other tweak I would make. Collecting messages from mapping should be the very last thing that happens otherwise we might lose something.

AudreyAltman replied:
:+1:

AudreyAltman commented:
@moltude @mdellabitta this is ready to be reviewed again.  Per conversation with Scott, I removed the runtime exception behavior from the `dplaUri` mapping as well, since it was related to the `originalId` mapping.  I also added a report of duplicate `originalId`s to the pretty-print mapping summary.  This now addresses tickets DT-760-4.  I tested it locally against a Digital Virginias harvest, which does not have any duplicate IDs.  It would be worth testing against Oklahoma or Getty, but I didn't have access to that harvest data.

moltude commented:
👍 after testing against Missouri

AudreyAltman commented:
This has been tested against Oklahoma and it looks ok.
